w3af/plugins/grep/analyze_cookies.py

```diff
@@ -187,6 +199,27 @@ def _analyze_cookie_security(self, request, response, cookie_obj,
         self._http_only(request, response, cookie_obj,
                         cookie_header_value, fingerprinted)
 
+    def _update_cookie_vulnerability(self, response, vulnerability):
+        # thread lock to escape race condition on db access
+        with self._plugin_lock:
+            vulns = kb.kb.get(self, 'security')
+        
+            if len(vulns) == 0:
+                kb.kb.append(self, 'security', vulnerability)
+            elif len(vulns) >= 1:
+                old_vuln = vulns[0]
```

Comment:
You were right on the method updating the vulnerability outside its type. I updated the method to now test on both the amount of vulnerabilities from the KB and the name of each returned vulnerability.

> The new URL (which is vulnerable to secure flag missing) is added to the first vulnerability, which is actually related to httponly

This is now fixed with:
https://github.com/andresriancho/w3af/pull/7763/files#diff-dceefea625bb1de5a6c0438ca6b36c2eR201
```
def _update_cookie_vulnerability(self, response, vulnerability):
 # thread lock to escape race condition on db access
 with self._plugin_lock:
 vulns = kb.kb.get(self, 'security')

 name = vulnerability.get_name()

 if len(vulns) == 0:
 kb.kb.append(self, 'security', vulnerability)
 else:
 for vuln in vulns:
 # Check if vuln and vulnerability are the same type
 if vuln.get_name() == name:
 old_vuln = vuln
 updated_vuln = copy.deepcopy(old_vuln)
 if 'urls' not in updated_vuln.keys():
 updated_vuln['urls'] = [old_vuln.get_url()]
 updated_vuln['urls'].append(response.get_url())
 kb.kb.update(old_vuln, updated_vuln)
 return updated_vuln
 # If no vulnerability of same type was found
 # append the vulnerability passed by caller
 kb.kb.append(self, 'security', vulnerability)
```
and tested with 
https://github.com/andresriancho/w3af/pull/7763/files#diff-36d002ea93080120a5535f35c1ed8417R344
```
def test_update_without_httponly_secure_over_https(self):
 body = ''
 url = URL('https://www.w3af.com/')
 headers = Headers({'content-type': 'text/html',
 'Set-Cookie': 'name="adf"'}.items())

 response = HTTPResponse(200, body, headers, url, url, _id=1)
 request = FuzzableRequest(url, method='GET')

 for i in range(0, 2):
 self.plugin.grep(request, response)

 self.assertEqual(len(kb.kb.get('analyze_cookies', 'cookies')), 1)
 self.assertEqual(len(kb.kb.get('analyze_cookies', 'security')), 2)
```